File: my-solutions/0023-merge-k-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def merge(self, list1, list2):
        head = ListNode(0)
        curr = head

        while(list1 and list2):
            if list1.val <= list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next

        if list1:
            curr.next = list1
        if list2:
            curr.next = list2

        return head.next
    
    def printList(self, list):
        while(list):
            if list.next == None:
                print(list.val, end = "")
            else:
                print(list.val, "->", end = "")
            list = list.next

        print("")
    
    def printAll(self, lists):
        for list in lists:
            self.printList(list)
        print("")

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def mergeLists(lists):
            if len(lists) <= 1:
                return lists
            size = len(lists)
            for i in range(size // 2):
                idx = 2 * i
                lists[i] = self.merge(lists[idx], lists[idx + 1])

            # self.printAll(lists)

            if size % 2 == 0:
                lists = lists[:size // 2]
            else:
                last = lists[-1]
                lists = lists[:size // 2]
                lists.append(last)

            # self.printAll(lists)
    
            return mergeLists(lists)
        
        res = mergeLists(lists)
        if res:
            return res[0]
        else:
            return None

File: my-solutions/0023-merge-k-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty()) return nullptr;
        return helper(lists, 0, lists.size() - 1);
    }
    ListNode* merge2Lists(ListNode* a, ListNode* b) {
        if(!a) return b;
        if(!b) return a;

        ListNode head(0);
        ListNode* ptr = &head;

        while(a && b) {
            if(a->val < b->val) {
                ptr->next = a;
                a = a->next;
            } else {
                ptr->next = b;
                b = b->next;
            }
            ptr = ptr->next;
        }

        ptr->next = a ? a : b;
        return head.next;
    }

    ListNode* helper(vector<ListNode*>& lists, int l, int r) {
        if(l == r) return lists[l];
        if(l + 1 == r) return merge2Lists(lists[l], lists[r]);
        int mid = (l + r) / 2;
        ListNode* left = helper(lists, l, mid);
        ListNode* right = helper(lists, mid + 1, r);
        return merge2Lists(left, right);
    }
};
